### PR TITLE
Disallow merge when required checked are missing

### LINKED ISSUE #21892: Show required commit status which haven’t been done yet
### Feature Description

If a base branch has a protection requiring a check, but the check hasn’t been started for a commit yet, then there is no indication in the pull which checks are missing. 

aside from pending, failed, succeeded etc checks, the pull page could also render missing checks (required checks that haven’t reported any status for the commit yet)

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by markusamshove**
fixes #21892 

This PR disallows merging a PR when not all commit status contexts configured in the branch protection are met.

Previously, the PR was happy to merge when one commit status was successful and the other contexts weren't reported.

Any feedback is welcome, first time Go :-)
I'm also not sure if the changes in the template break something else

Given the following branch protection:

branch_protection

This was shown before the change:

before

With the change, it is now shown as this:

after

### CODE REVIEW

File: routers/web/repo/pull.go [hunk 1]
```diff
@@ -652,6 +652,24 @@ func PrepareViewPullInfo(ctx *context.Context, issue *issues_model.Issue) *git.C
 	}
 
 	if pb != nil && pb.EnableStatusCheck {
+
+		missingRequiredChecks := make([]string, 0)
```
Comment (wxiaoguang) [inline routers/web/repo/pull.go]:
```suggestion
		var missingRequiredChecks []string
```

**Review by wxiaoguang — APPROVED** [review]

**Review by delvh — APPROVED** [review]

File: routers/web/repo/pull.go [hunk 2]
```diff
@@ -652,6 +652,24 @@ func PrepareViewPullInfo(ctx *context.Context, issue *issues_model.Issue) *git.C
 	}
 
 	if pb != nil && pb.EnableStatusCheck {
+
+		var missingRequiredChecks []string
+		for _, requiredContext := range pb.StatusCheckContexts {
+			contextFound := false
+			for _, presentStatus := range commitStatuses {
+				if presentStatus.Context == requiredContext {
```
Comment (lunny) [inline routers/web/repo/pull.go]: `requiredContext` maybe a glob express, I think it should also be handled here?

File: routers/web/repo/pull.go [hunk 3]
```diff
@@ -652,6 +652,24 @@ func PrepareViewPullInfo(ctx *context.Context, issue *issues_model.Issue) *git.C
...
+				}
+			}
+
+			if !contextFound {
+				missingRequiredChecks = append(missingRequiredChecks, requiredContext)
+			}
+		}
+		ctx.Data["MissingRequiredChecks"] = missingRequiredChecks
+
+		ctx.Data["RequiredStatusChecks"] = pb.StatusCheckContexts
```
Comment (wxiaoguang) [inline routers/web/repo/pull.go]:
```suggestion
```

By the way, is this `RequiredStatusChecks` used anywhere? If no, it could be removed.

**Review by lunny — DISMISSED** [review]
per https://github.com/go-gitea/gitea/pull/29143/files#r1493155168

File: routers/web/repo/pull.go (hunk 3 above)
Comment (markusamshove) [reply routers/web/repo/pull.go]: I grepped through the code and this is the only place where ` RequiredStatusChecks` is present, so I removed it :+1:

**Comment by MarkusAmshove** [comment]
> per https://github.com/go-gitea/gitea/pull/29143/files#r1493155168

Done

**Review by Zettat123 — APPROVED** [review]

**Pull request merged**